src/main/java/org/testng/internal/Parameters.java

```diff
@@ -233,19 +260,27 @@ private static void checkParameterTypes(String methodName,
       }
     }
 
-    if ( (parameterNames.length ==0) && (totalLength != 0) ) {
+
+    if ((parameterNames.length == 0)) {
       //parameterNames is usually populated via the @Parameters annotation, so we would need to
       //apply our logic only when @Parameters annotation is not involved.
-      String errPrefix = "Cannot inject " + methodAnnotation + " annotated Method [" + methodName + "] with "
-            + Arrays.toString(parameterTypes);
-      if (mapping.containsKey(methodAnnotation)) {
-        errPrefix = "Can inject only one of " + prettyFormat(mapping.get(methodAnnotation)) +
-            " into a " + methodAnnotation + " annotated " + methodName;
+      String annotation = methodAnnotation;
+      if (!methodAnnotation.startsWith("@")) {
+        annotation = "@" + methodAnnotation;
+      }
+      boolean invalid = (totalLength != 0) || (!validParameters(methodAnnotation, parameterTypes));
+      if (invalid) {
+        String errPrefix = "Cannot inject " + annotation + " annotated Method [" + methodName + "] with "
```

Comment:
In fact, the idea was to move the `errPrefix` init in the if/else block. Could you just change it and squash your commits before the merge?